File: src/searchat/agents/registry.py
"""Registry for agent-specific conversation providers."""

from __future__ import annotations

from pathlib import Path
from typing import Iterable, Optional

from .base import AgentProvider
from .claude import ClaudeProvider
from .codex import CodexProvider
from .vibe import VibeProvider


_PROVIDERS = (
    CodexProvider(),
    ClaudeProvider(),
    VibeProvider(),
)


def iter_providers() -> Iterable[AgentProvider]:
    """Return all registered providers in match priority order."""
    return _PROVIDERS


def get_provider(agent_id: str) -> AgentProvider:
    """Return a provider by id."""
    for provider in _PROVIDERS:
        if provider.agent_id == agent_id:
            return provider
    raise KeyError(f"Unknown agent provider: {agent_id}")


def detect_provider(file_path: Path) -> Optional[AgentProvider]:
    """Return the provider responsible for a transcript file."""
    for provider in _PROVIDERS:
        if provider.matches_file(file_path):
            return provider
    return None


def detect_provider_id(file_path: str) -> Optional[str]:
    """Convenience wrapper for string file paths."""
    provider = detect_provider(Path(file_path))
    return provider.agent_id if provider else None

File: rust/crates/searchat-agents/src/registry.rs
//! Registry of all known agent providers.
//!
//! Match priority follows the Python registry order: Codex > Claude > Vibe.
//! Codex files (`.jsonl` under `/.codex/`) must be checked before Claude
//! because both use the `.jsonl` extension.

use std::path::{Path, PathBuf};

use crate::{AgentProvider, ClaudeProvider, CodexProvider, VibeProvider};

/// Return one instance of every registered provider in match-priority order.
///
/// Priority: Codex → Claude → Vibe
pub fn all_providers() -> Vec<Box<dyn AgentProvider>> {
    vec![
        Box::new(CodexProvider),
        Box::new(ClaudeProvider),
        Box::new(VibeProvider),
    ]
}

/// Return the provider that claims `path`, or `None` if no provider matches.
///
/// Providers are checked in priority order (Codex > Claude > Vibe).
pub fn detect_provider(path: &Path) -> Option<Box<dyn AgentProvider>> {
    for provider in all_providers() {
        if provider.matches_file(path) {
            return Some(provider);
        }
    }
    None
}

/// Return all discovered `(provider_id, directory)` pairs across every provider.
///
/// Directories that don't exist on the current machine are included for Codex
/// WSL paths (they may still be walkable) but skipped for others.
pub fn iter_dirs() -> Vec<(String, PathBuf)> {
    all_providers()
        .into_iter()
        .flat_map(|p| {
            let id = p.id().to_owned();
            p.discover_dirs().into_iter().map(move |dir| (id.clone(), dir))
        })
        .collect()
}
